src/client/oicp/CpoOICPClient.ts

```diff
@@ -0,0 +1,1303 @@
+        action: ServerAction.OICP_PUSH_CDRS,
+        message: `OICP data does not exists on Session ID '${transaction.id}'`,
+        module: MODULE_NAME, method: 'pushCdr',
+      });
+    }
+    if (!transaction.oicpData.session) {
+      throw new BackendError({
+        source: transaction.chargeBoxID,
+        action: ServerAction.OICP_PUSH_CDRS,
+        message: `OICP Session data does not exists on Session ID '${transaction.id}'`,
+        module: MODULE_NAME, method: 'pushCdr',
+      });
+    }
+    if (!transaction.stop) {
+      throw new BackendError({
+        source: transaction.chargeBoxID,
+        action: ServerAction.OICP_PUSH_CDRS,
+        message: `OICP Session ID '${transaction.oicpData.session.id}' (ID '${transaction.id}') not stopped`,
+        module: MODULE_NAME, method: 'pushCdr',
+      });
+    }
+
+    // Get CDR endpoint url
+    // for QA - environment: https://service-qa.hubject.com/api/oicp/cdrmgmt/v22/operators/{operatorID}/charge-detail-record
+    // for PROD - environment: https://service.hubject.com/api/oicp/cdrmgmt/v22/operators/{operatorID}/charge-detail-record
+    const fullUrl = this.getEndpointUrl('cdrs', ServerAction.OICP_PUSH_CDRS);
+
+    const cdr: OICPChargeDetailRecord = {} as OICPChargeDetailRecord;
+    cdr.SessionID = transaction.oicpData.session.id;
+    cdr.CPOPartnerSessionID = String(transaction.id); // Optional
+    cdr.EMPPartnerSessionID = transaction.oicpData.session.empPartnerSessionID; // Optional
+    cdr.EvseID = transaction.oicpData.session.evse.EvseID;
+    cdr.Identification = transaction.oicpData.session.identification;
+    cdr.ChargingStart = transaction.timestamp;
+    cdr.ChargingEnd = transaction.stop.timestamp;
+    cdr.SessionStart = transaction.oicpData.session.start_datetime;
+    cdr.SessionEnd = transaction.oicpData.session.end_datetime;
+    cdr.MeterValueStart = parseFloat((transaction.meterStart / 1000).toFixed(3)); // Optional
+    cdr.MeterValueEnd = parseFloat((transaction.stop.meterStop / 1000).toFixed(3)); // Optional
+    cdr.MeterValueInBetween = transaction.oicpData.session.meterValueInBetween.map((mV) => parseFloat((mV / 1000).toFixed(3))); // Optional
+    cdr.ConsumedEnergy = parseFloat((transaction.stop.totalConsumptionWh / 1000).toFixed(3)); // In kW.h
+    cdr.SignedMeteringValues; // Optional
+    cdr.CalibrationLawVerificationInfo; // Optional
+    cdr.HubOperatorID; // Optional
+    cdr.HubProviderID; // Optional
+
+    transaction.oicpData.cdr = cdr;
+    const payload: OICPChargeDetailRecord = transaction.oicpData.cdr;
+
+    // Log
+    Logging.logDebug({
+      tenantID: this.tenant.id,
+      action: ServerAction.OICP_PUSH_CDRS,
+      message: `Post CDR of OICP Transaction ID '${transaction.oicpData.session.id}' (ID '${transaction.id}') at ${fullUrl}`,
+      module: MODULE_NAME, method: 'pushCdr',
+      detailedMessages: { payload: transaction.oicpData.cdr }
+    });
+
+    const httpsAgent = new https.Agent({
```

Comment:
You can factorize the httpsAgent creation in an helper method and use it everywhere.

Reply:
Okay, I did it